alpenhorn/Task.py

```diff
@@ -0,0 +1,1021 @@
+        del_files = ar.ArchiveFileCopy.select().where(
+            dfclause,
+            ar.ArchiveFileCopy.node == self.node,
+            ar.ArchiveFileCopy.has_file == "Y",
+        )
+
+        # Process candidates for deletion
+        del_count = 0  # Counter for no. of deletions (limits no. per node update)
+        for fcopy in del_files.order_by(ar.ArchiveFileCopy.id):
+
+            # Limit number of deletions to 500 per main loop iteration.
+            if del_count >= 500:
+                break
+
+            # Get all the *other* copies.
+            other_copies = fcopy.file.copies.where(
+                ar.ArchiveFileCopy.id != fcopy.id, ar.ArchiveFileCopy.has_file == "Y"
+            )
+
+            # Get the number of copies on archive nodes
+            ncopies = (
+                other_copies.join(st.StorageNode)
+                .where(st.StorageNode.storage_type == "A")
+                .count()
+            )
+
+            shortname = "%s/%s" % (fcopy.file.acq.name, fcopy.file.name)
+            fullpath = "%s/%s/%s" % (
+                self.node.root,
+                fcopy.file.acq.name,
+                fcopy.file.name,
+            )
+
+            # If at least two other copies we can delete the file.
+            if ncopies >= 2:
+
+                # Use transaction such that errors thrown in the os.remove do not leave
+                # the database inconsistent.
+                with db.database_proxy.transaction():
+                    if os.path.exists(fullpath):
+                        os.remove(fullpath)  # Remove the actual file
+
+                        # Check if the acquisition directory or containing directories are now empty,
+                        # and remove if they are.
+                        dirname = os.path.dirname(fullpath)
+                        while dirname != self.node.root:
+                            if not os.listdir(dirname):
+                                log.info(
+                                    "Removing acquisition directory %s on %s"
+                                    % (fcopy.file.acq.name, fcopy.node.name)
+                                )
+                                os.rmdir(dirname)
+                                dirname = os.path.dirname(dirname)
+                            else:
+                                break
+
+                    fcopy.has_file = "N"
+                    fcopy.wants_file = "N"  # Set in case it was 'M' before
+                    fcopy.save()  # Update the FileCopy in the database
+
+                    log.info("Removed file copy: %s" % shortname)
+
+                del_count += 1
+
+            else:
+                log.info("Too few backups to delete %s" % shortname)
+
+
+class DiskTransferTask(Task):
+    def run(self):
+        """Process file copy requests onto this node."""
+
+        print("{} run()...".format(type(self).__name__))
+
+        global done_transport_this_cycle
```

Comment:
I do not think it does. At minimum, we will need to import `done_transport_this_cycle`. But even with that, I am not totally sure if the `global` keyword allows the setting of globals from another file, I am going to run a test.

Reply:
But even if it works, we are probably better off doing something else than this.

Reply:
(The `global` keyword does allow the setting of globals from another file.)

Reply:
Yup. 

I think done_transport_this_cycle is only really necessary on the DiskTransfer tasks, so it could be moved to a class variable instead. That's definitely a lot cleaner.

There might still be a better way of doing this though. I'm always a little unclear of what the point of it is. I *think* it's to make sure that transport disk copies end up filling up each disk in turn rather than spreading out over all the available disks.